plugin/pkg/admission/webhook/admission.go

```diff
@@ -0,0 +1,186 @@
+		),
+	}, nil
+}
+
+type HookSource interface {
+	List() ([]apiserver.ExternalAdmissionHook, error)
+}
+
+type ServiceResolver interface {
+	ResolveEndpoint(namespace, name string) (*url.URL, error)
+}
+
+// GenericAdmissionWebhook is an implementation of admission.Interface.
+type GenericAdmissionWebhook struct {
+	*admission.Handler
+	hookSource      HookSource
+	serviceResolver ServiceResolver
+}
+
+// Admit makes an admission decision based on the request attributes.
+func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) (err error) {
+	hooks, err := a.hookSource.List()
+	if err != nil {
+		return fmt.Errorf("failed listing hooks: %v", err)
+	}
+
+	var errCount int64
+	var failCount int64
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(hooks))
+	for i := range hooks {
+		go func(hook *apiserver.ExternalAdmissionHook) {
+			defer wg.Done()
+			allowed, err := a.callHook(hook, attr)
+			if err != nil {
+				atomic.AddInt64(&errCount, 1)
+				glog.Warningf("Failed calling webhook %v: %v", hook.Name, err)
+			} else if !allowed {
+				atomic.AddInt64(&failCount, 1)
+			}
+		}(&hooks[i])
+	}
+	wg.Wait()
+
+	if failCount > 0 {
+		return ErrWebhookRejected
+	}
+
+	// TODO: implement fail-closed by checking if errCount > 0
+	return nil
+}
+
+func (a *GenericAdmissionWebhook) callHook(h *apiserver.ExternalAdmissionHook, attr admission.Attributes) (allow bool, err error) {
+	matches := false
+	for _, r := range h.Rules {
+		m := RuleMatcher{Rule: r, Attr: attr}
+		if m.Matches() {
+			matches = true
+			break
+		}
+	}
+	if !matches {
+		return true, nil
+	}
+
+	// Make the webhook request
+	request := admissionv1alpha1.NewAdmissionReview(attr)
+	client, err := a.hookClient(h)
+	if err != nil {
+		return false, fmt.Errorf("unable to construct hook %q client: %v", h.Name, err)
+	}
+	response := client.Post().Body(&request).Do()
+
+	// Handle webhook response
+	if err := response.Error(); err != nil {
+		return a.handleError(attr, matched.FailAction, err)
+	}
+
+	var statusCode int
+	if response.StatusCode(&statusCode); statusCode < 200 || statusCode >= 300 {
+		return false, fmt.Errorf("error contacting webhook %q: %d", h.Name, statusCode)
+	}
+
+	if err := response.Into(&request); err != nil {
+		return false, fmt.Errorf("error interpreting webhook %q response: %v", h.Name, err)
+	}
+
+	if !request.Status.Allowed {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (a *GenericAdmissionWebhook) hookClient(h *apiserver.ExternalAdmissionHook) (*rest.RESTClient, error) {
+	u, err := a.serviceResolver.ResolveEndpoint(h.ClientConfig.Service.Namespace, h.ClientConfig.Service.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: cache these instead of constructing one each time
+	cfg := rest.Config{
+		Host:    u.Host,
+		APIPath: u.Path,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: h.ClientConfig.CABundle,
+			// TODO: set KeyData/CertData
+		},
+		UserAgent: "kube-apiserver-admission",
+		Timeout:   30 * time.Second,
```

Comment:
Any reason we cannot make this configurable?  Might not be necessary for 1.0 but I can see this being useful in the future.